Adding Low/High side flag

shubham3121 opened the pull request:
## Description
Please include a summary of the change, the motivation, and any additional context that will help others understand your PR. If it closes one or more open issues, [please tag them as described here](https://help.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue#linking-a-pull-request-to-an-issue-using-a-keyword).

- [x] add high and low side flags to:
     - [x] orchestra
     - [x] settings
     - [x] metadata
     - [x] hagrid
- [x] Create SyftWarning class which can propogated from server to client:
- [x] upon logging -> show node side type -> low/high
- [x] upon dataset upload and list for the low side, warning "This is the low-side domain, do not upload private assets."
- [x] register
- [x] toggle for user signup

## Affected Dependencies
List any dependencies that are required for this change.



## How has this been tested?
- Describe the tests that you ran to verify your changes.
- Provide instructions so we can reproduce.
- List any relevant details for your test configuration.

## Checklist
- [ ] I have followed the [Contribution Guidelines](https://github.com/OpenMined/.github/blob/master/CONTRIBUTING.md) and [Code of Conduct](https://github.com/OpenMined/.github/blob/master/CODE_OF_CONDUCT.md)
- [ ] I have commented my code following the [OpenMined Styleguide](https://github.com/OpenMined/.github/blob/master/STYLEGUIDE.md)
- [ ] I have labeled this PR with the relevant [Type labels](https://github.com/OpenMined/.github/labels?q=Type%3A)
- [ ] My changes are covered by tests

shubham3121 committed:
add high and low side flags to orchestra, settings, metadata, hagrid

shubham3121 committed:
fix passing node side type to NodeHandle

shubham3121 committed:
add _repr_html_ to both gateway and enclave client

shubham3121 committed:
add repr html to all client types

shubham3121 committed:
add warning on low side node during dataset upload
add node_side_type information to notebook

shubham3121 committed:
add flag to enable/disable prompts
add a method to print warning prompts with confirmations
add warning/reminder messages to dataset, connect_to_gateway

shubham3121 committed:
disable prompt flag during pytest

shubham3121 committed:
pass node side type to uvicorn server

shubham3121 committed:
add prompt on registeration for high side
add prompt on connect to gateway at EnclaveClient

shubham3121 committed:
fix import

shubham3121 committed:
add node side type to WorkerSettings
update notebook

shubham3121 commented:
Apply to gateway (If node is a high side node):
![Screenshot from 2023-07-17 17-49-57](https://github.com/OpenMined/PySyft/assets/11032835/6978d828-f124-4d69-97db-ab7492c894df)

Only warning (low side node -> no confirmation)
![Screenshot from 2023-07-17 17-49-31](https://github.com/OpenMined/PySyft/assets/11032835/d7d5481c-852e-4feb-a519-1201000725f8)

Upload Dataset (Only High side):
High Side
![Screenshot from 2023-07-17 17-50-51](https://github.com/OpenMined/PySyft/assets/11032835/569556d5-ad09-4ad9-8f28-f8de9830d351)

User Registration (High Side Only) -> This is shown only if guest signup is disabled and node is High Side
![Screenshot from 2023-07-17 17-51-31](https://github.com/OpenMined/PySyft/assets/11032835/b6a7e752-7c7e-4f39-a5d8-6f0d07e23fc4)

Get Dataset (High and Low Side) -> Only Reminder
![Screenshot from 2023-07-17 17-55-56](https://github.com/OpenMined/PySyft/assets/11032835/a41377d4-6133-49c1-8342-4ea2e6b32ba7)

![Screenshot from 2023-07-14 18-46-58](https://github.com/OpenMined/PySyft/assets/11032835/4664f7ae-dd7a-4a78-87d0-33325ac6760b)

Toggle Registration (Only High Side):
![Screenshot from 2023-07-17 17-56-38](https://github.com/OpenMined/PySyft/assets/11032835/07a85c26-7b29-471b-b2ea-773a9e653263)


Login Screens:

On login by a user:
![Screenshot from 2023-07-12 18-48-09](https://github.com/OpenMined/PySyft/assets/11032835/b9db3041-b447-47be-a2f1-3a3f4f56cdf7)

![Screenshot from 2023-07-12 18-48-42](https://github.com/OpenMined/PySyft/assets/11032835/b27ef80c-d16c-4f1e-9e99-d68fbc1ef3af)

Client  Repr: Added Node Type and Node Side Type information 
- Gateway: 
![Screenshot from 2023-07-12 20-08-29](https://github.com/OpenMined/PySyft/assets/11032835/373ff140-cb55-4f17-b7f5-3dec48079b7e)

- Domain: 
![Screenshot from 2023-07-12 20-07-13](https://github.com/OpenMined/PySyft/assets/11032835/0cc28849-c66f-4d67-8701-e8f2cd94e4cf)
- Enclaves:
![Screenshot from 2023-07-12 20-06-29](https://github.com/OpenMined/PySyft/assets/11032835/45d35631-0b10-4fe8-8a70-cecedf2ac72f)

koenvanderveen commented on notebooks/tutorials/enclaves/Enclave-single-notebook-high-low-network.ipynb:
```diff
@@ -13,6 +13,17 @@
     "from recordlinkage.datasets import load_febrl4"
```
<pre class="ql-syntax" spellcheck="false">Line #2.    it_node_low = sy.Orchestra.launch(name="italy-1", node_side_type="low", local_db=True, reset=True) 
</pre><p>do we want to use high/low here or staging/production? Also, could be helpful to make it a boolean as its easier to autocomplete? is_staging?</p>
<br/>

 _Reply via <a href='https://app.reviewnb.com/OpenMined/PySyft/pull/7916/discussion/#comment-1261334417'>ReviewNB</a>_ <div id='ReviewNBCommentContext-DoNotDelete' style='display:none' data-state='OPEN' data-cellIndex='7'/>

shubham3121 replied:
For now we're using high/low on client side, we can switch to prod/staging if that's required.

koenvanderveen commented on packages/syft/src/syft/client/domain_client.py:
```diff
@@ -60,6 +62,21 @@ def upload_dataset(self, dataset: CreateDataset) -> Union[SyftSuccess, SyftError
+                "This is an operation on the high side domain, "
+                "and the datasets could contain private information."
+            )
+        if not prompt_warning_message(message=message, confirm=True):
```
I think we only have to prompt in the case of the low side, as that is very risky. The high side could just be an info based on discussion with @IrinaMBejan as there is less that could go wrong

koenvanderveen commented:
would be good to also have a prompt for when user registration is enabled on the high side

koenvanderveen commented:
How does it work now with the notebooks that dont have a high side and low side. Bcs the default node type is high so by default you get prompts everywhere? Should be turned off for notebook testing I suppose. Should maybe also be turned off for the tutorials?

shubham3121 committed:
add SyftWarning
add APIEndpointWarning and WarningContext

shubham3121 committed:
integrate CRUD warning

shubham3121 committed:
- create different types of warnings
- add APIEndpointWarning types to dataset, register enable, dataset upload

shubham3121 committed:
fix node side type check in HighSideWarning

shubham3121 commented:
https://github.com/OpenMined/PySyft/pull/7916#issuecomment-1632757831
Flag is turned off for tests.

> would be good to also have a prompt for when user registration is enabled on the high side

Added to toggle registration api as well

shubham3121 committed:
fix lint

shubham3121 committed:
fix node side type passed twice

shubham3121 committed:
add custom repr to APIEndpointWarning

shubham3121 committed:
- return a new class instance on creating a message from context
- fix tests

shubham3121 committed:
enable warning via hagrid
skip warnings if flag is disabled

shubham3121 committed:
fix warnings flag in hagrid cli

shubham3121 committed:
- add enable_warnings flag to Node level
- propogate enable warnings flag via Orchestra
- add warning enable flag to APIEndpointWarning object itself
- remove flags.WARNING_ENABLED

shubham3121 committed:
fix minor bugs

shubham3121 committed:
import node type from syft in orchestra

shubham3121 committed:
fix notebook

shubham3121 committed:
prompt warning on request approve in case of code is on enclaves

shubham3121 committed:
Merge branch 'dev' into low-high-side-flag

shubham3121 committed:
make low side / high side its own flag

shubham3121 committed:
fix parsing of low side flag

rasswanth-s commented on packages/grid/default.env:
```diff
@@ -95,6 +95,7 @@ JAEGER_PORT=14268
 
 # Syft
 SYFT_TUTORIAL_MODE=False
+ENABLE_WARNINGS=True
```
Q: Since by orchestra , by default , enable warnings is False, should we set here also to False

shubham3121 replied:
Made it false for orchestra by default because its used in test and notebooks, where we want it to be false

shubham3121 committed:
Merge branch 'dev' into low-high-side-flag

rasswanth-s commented on packages/hagrid/hagrid/cli.py:
```diff
@@ -426,6 +432,16 @@ def clean(location: str) -> None:
     help="Render Docker Files",
 )
+@click.option(
+    "--no-warnings",
```
Q: Since, by default, we thought of no warnings, we  could re-name it to --enable-warnings

rasswanth-s commented on packages/syft/src/syft/service/user/user_warning.py:
```diff
@@ -0,0 +1 @@
+
```
Q: Should  we delete this file as it is empty?

shubham3121 replied:
Oh right !! added by mistake 😄

shubham3121 committed:
delete unused file

rasswanth-s commented on packages/syft/src/syft/client/client.py:
```diff
@@ -768,7 +782,10 @@ def login(
         if verbose:
-            print(f"Logged into {_client.name} as GUEST")
+            print(
+                f"Logged into {_client.name} {_client.metadata.node_side_type} "
```
Could we have separation with angle brackets, to make it clear

like
"Logged into < {_client.name} - {_client.metadata.node_type} > - {_client.metadata.node_side_type} side as GUEST"

or any other nice way to make it coherent.

rasswanth-s replied:
When I tried it , it was hard to distinguish without separation.

shubham3121 replied:
fixed: https://github.com/OpenMined/PySyft/pull/7916/commits/e3dbf0b80c65ebe49c6791faad1b542d73ba9381

rasswanth-s commented on packages/syft/src/syft/client/client.py:
```diff
@@ -588,7 +587,10 @@ def login(
             # self.__user_role =
             self._fetch_api(self.credentials)
-            print(f"Logged into {self.name} as <{email}>")
+            print(
```
Same comment: https://github.com/OpenMined/PySyft/pull/7916/files#r1265269590

shubham3121 replied:
Fixed: https://github.com/OpenMined/PySyft/pull/7916/commits/e3dbf0b80c65ebe49c6791faad1b542d73ba9381

shubham3121 commented:
Now everything is `SyftWarning` instead of `CRUDWarning` or `HighSideCrudWarning`, etc. some images are old

shubham3121 committed:
update message repr on login
remove crud warning from request apply service method

rasswanth-s commented on packages/syft/src/syft/client/enclave_client.py:
```diff
@@ -126,3 +129,73 @@ def request_code_execution(self, code: SubmitUserCode):
 
         return res
+
+    def _repr_html_(self) -> str:
```
Well Done ❤️

rasswanth-s commented on packages/syft/src/syft/service/request/request.py:
```diff
@@ -271,6 +272,18 @@ def approve(self):
+        if code and not isinstance(code, SyftError):
+            is_enclave = getattr(code, "enclave_metadata", None) is not None
+
+        if is_enclave:
```
STRETCH: Maybe if it is  not an enclave (have an else), we display the high side , low side warning on the client side as we removed it from the request service side.

as the current request does not raise a prompt if the request is not of type enclave.

shubham3121 replied:
Fixed by: https://github.com/OpenMined/PySyft/pull/7916/commits/85569e72674e2b015af149cf24b471dfa785ca38

rasswanth-s reviewed (approved):
Awesome Work @shubham3121  🔥  🔥 , two minor comments(https://github.com/OpenMined/PySyft/pull/7916#discussion_r1265349686,

https://github.com/OpenMined/PySyft/pull/7916#discussion_r1265375115 )

rasswanth-s commented:
Great Functionality in a short amount of time.
Kachow
![lightning-mcqueen-cars-movie](https://github.com/OpenMined/PySyft/assets/43314053/74e8a08c-4391-4f34-ade7-5ecdd2dd00bc)

rasswanth-s committed:
Merge branch 'dev' into low-high-side-flag

rasswanth-s commented on packages/grid/backend/grid/core/node.py:
```diff
@@ -27,15 +29,23 @@
 else:
     worker = Domain(
         name=node_name,
+        node_side_type=node_side_type,
```
We could add also in worker.py, as that is being used for the single container deployments.

rasswanth-s replied:
packages/grid/worker/worker.py

shubham3121 replied:
Fixed by: https://github.com/OpenMined/PySyft/pull/7916/commits/85569e72674e2b015af149cf24b471dfa785ca38

shubham3121 committed:
- set low side type and enable warning vars to Worker class
- pass low/high side and enable warnings flag to worker compose file

koenvanderveen commented:
this looks very good, no further comments
